File: src/lib/pjecalc/core/comum/periodo-de-juros.ts
/**
 * PJe-Calc v2.15.1 — PeriodoDeJuros
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.comum.PeriodoDeJuros
 *
 * Ref Java: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/comum/PeriodoDeJuros.java
 *
 * Representa um segmento de juros com aliquota, tipoQuantidade (INTEIRO ou FRACAO),
 * tipoJuros (SIMPLES ou COMPOSTOS) e tabelaJuros (JurosEnum). A propriedade
 * crítica é `getMeses()` — contagem pro-rata die nos extremos quando FRACAO.
 */
import Decimal from 'decimal.js';
import { HelperDate } from '../base/comum/helper-date';
import { multiplicar, nulo } from '../base/comum/utils';
import { JurosEnum, TipoDeJurosEnum, TipoDeQuantidadeDeJurosBaseEnum } from '../constantes/enums';

export class PeriodoDeJuros {
  private dataInicial: Date | null = null;
  private dataFinal: Date | null = null;
  private aliquota: Decimal | null = null;
  private tipoDeQuantidade: TipoDeQuantidadeDeJurosBaseEnum | null = null;
  private tipoDeJuros: TipoDeJurosEnum | null = null;
  private proximoPeriodo: PeriodoDeJuros | null = null;
  private fazendaPublica: boolean = false;
  private tabelaJuros: JurosEnum | null = null;
  private meses: Decimal | null = null;
  private taxa: Decimal | null = null;

  constructor(
    dataInicial?: Date | null,
    dataFinal?: Date | null,
    aliquota?: Decimal | null,
    tipoDeQuantidade?: TipoDeQuantidadeDeJurosBaseEnum | null,
    tipoDeJuros?: TipoDeJurosEnum | null,
    fazendaPublica?: boolean,
    tabelaJuros?: JurosEnum | null
  ) {
    if (dataInicial !== undefined) this.dataInicial = dataInicial ?? null;
    if (dataFinal !== undefined) this.dataFinal = dataFinal ?? null;
    if (aliquota !== undefined) this.aliquota = aliquota ?? null;
    if (tipoDeQuantidade !== undefined) this.tipoDeQuantidade = tipoDeQuantidade ?? null;
    if (tipoDeJuros !== undefined) this.tipoDeJuros = tipoDeJuros ?? null;
    if (fazendaPublica !== undefined) this.fazendaPublica = fazendaPublica;
    if (tabelaJuros !== undefined) this.tabelaJuros = tabelaJuros ?? null;
  }

  clone(): PeriodoDeJuros {
    return new PeriodoDeJuros(
      this.dataInicial, this.dataFinal, this.aliquota,
      this.tipoDeQuantidade, this.tipoDeJuros, this.fazendaPublica, this.tabelaJuros
    );
  }

  getDataInicial(): Date | null { return this.dataInicial; }
  setDataInicial(d: Date): void { this.dataInicial = d; }
  getDataFinal(): Date | null { return this.dataFinal; }
  setDataFinal(d: Date): void { this.dataFinal = d; }

  getAliquota(): Decimal | null { return this.aliquota; }
  setAliquota(a: Decimal): void { this.aliquota = a; }

  getTipoDeQuantidade(): TipoDeQuantidadeDeJurosBaseEnum | null { return this.tipoDeQuantidade; }
  setTipoDeQuantidade(t: TipoDeQuantidadeDeJurosBaseEnum): void { this.tipoDeQuantidade = t; }

  getTipoDeJuros(): TipoDeJurosEnum | null { return this.tipoDeJuros; }
  setTipoDeJuros(t: TipoDeJurosEnum): void { this.tipoDeJuros = t; }

  getProximoPeriodo(): PeriodoDeJuros | null { return this.proximoPeriodo; }
  setProximoPeriodo(p: PeriodoDeJuros | null): void { this.proximoPeriodo = p; }

  isFazendaPublica(): boolean { return this.fazendaPublica; }
  setFazendaPublica(v: boolean): void { this.fazendaPublica = v; }

  getTabelaJuros(): JurosEnum | null { return this.tabelaJuros; }
  setTabelaJuros(t: JurosEnum): void { this.tabelaJuros = t; }

  /** getDias (linha 94) — dias inclusive */
  getDias(): number {
    if (!this.dataInicial || !this.dataFinal) return 0;
    return HelperDate.countDays(this.dataInicial, this.dataFinal) + 1;
  }

  /**
   * getMeses (linha 98) — CORE da paridade de juros.
   *
   * Contagem de meses com FRAÇÃO pro-rata die quando tipoDeQuantidade = FRACAO.
   *
   * Algoritmo (linhas 98-125 do Java):
   *   meses_inteiros = countMonths(dataInicial, dataFinal)
   *   diasRestantesNoMesInicial = countDays(dataInicial, lastDayOfMonth) + 1
   *
   *   Se FRACAO:
   *     se diasRestantes_mesInicial < diasNoMesInicial:
   *       meses = meses - 1 + (diasRestantes / diasNoMes)
   *     se diaFinal < diasNoMesFinal:
   *       meses = meses - 1 + (diaFinal / diasNoMes)
   *
   *   Se INTEIRO:
   *     se diasRestantes < 15: meses -= 1
   *     se diaFinal < 15: meses -= 1
   */
  getMeses(): Decimal {
    if (this.meses === null) {
      if (!this.dataInicial || !this.dataFinal) return new Decimal(0);
      const hdInicial = HelperDate.getInstance(this.dataInicial)!;
      const hdFinal = HelperDate.getInstance(this.dataFinal)!;
      this.meses = new Decimal(HelperDate.countMonths(this.dataInicial, this.dataFinal));
      const quantDiasInicial = HelperDate.countDays(this.dataInicial, hdInicial.lastDayOfTheMonth().getDate()) + 1;

      if (this.tipoDeQuantidade === TipoDeQuantidadeDeJurosBaseEnum.FRACAO) {
        const totalDiasMesInicial = hdInicial.daysInMonth();
        if (quantDiasInicial < totalDiasMesInicial) {
          const fracao = new Decimal(quantDiasInicial).div(totalDiasMesInicial);
          this.meses = this.meses.minus(1).plus(fracao);
        }
        const diaFinal = hdFinal.getDay();
        const totalDiasMesFinal = hdFinal.daysInMonth();
        if (diaFinal < totalDiasMesFinal) {
          const fracao = new Decimal(diaFinal).div(totalDiasMesFinal);
          this.meses = this.meses.minus(1).plus(fracao);
        }
      } else {
        // INTEIRO — arredonda ao 15º dia
        if (quantDiasInicial < 15) this.meses = this.meses.minus(1);
        if (hdFinal.getDay() < 15) this.meses = this.meses.minus(1);
      }
    }
    return this.meses;
  }

  /**
   * getTaxa (linha 127) — retorna a TAXA PERCENTUAL do período.
   * - Para JUROS_ZERO_TRINTA_TRES (0,0333% a.d.): taxa = aliquota × dias
   * - Para SIMPLES: taxa = aliquota × meses_fracionados
   * - Para COMPOSTOS: taxa = ((1 + 0,01)^meses − 1) × 100 (usa aliquota=1%)
   *
   * NOTA: composto do PJe-Calc assume 1%/mês fixo (não usa aliquota).
   * Retorna TAXA em % (ex: 13.5 para 13,5%), não fator.
   */
  getTaxa(): Decimal {
    if (this.taxa === null) {
      if (this.tabelaJuros === JurosEnum.JUROS_ZERO_TRINTA_TRES) {
        this.taxa = multiplicar(this.getAliquota(), new Decimal(this.getDias()))!;
      } else if (this.tipoDeJuros === TipoDeJurosEnum.SIMPLES) {
        this.taxa = this.getAliquota()!.times(this.getMeses());
      } else {
        // Composto: usa 1% fixo e calcula (1.01)^meses
        const base = new Decimal(1).div(100).plus(1); // 1.01
        this.taxa = base.pow(this.getMeses()).minus(1).times(100);
      }
    }
    return this.taxa;
  }

  toString(): string {
    const fmt = (d: Date | null) => d ? HelperDate.getInstance(d)!.format('dd/MM/yyyy') : '';
    return `[periodo: ${fmt(this.dataInicial)} - ${fmt(this.dataFinal)}, meses: ${this.getMeses().toString()}, taxa: ${this.getTaxa().toString()}]`;
  }
}

File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/comum/PeriodoDeJuros.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.comum;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.constantes.JurosEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoDeJurosEnum;
import br.jus.trt8.pjecalc.negocio.constantes.TipoDeQuantidadeDeJurosBaseEnum;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.Date;

public class PeriodoDeJuros
implements Serializable {
    private static final long serialVersionUID = 1L;
    private Date dataInicial;
    private Date dataFinal;
    private BigDecimal aliquota;
    private TipoDeQuantidadeDeJurosBaseEnum tipoDeQuantidade;
    private TipoDeJurosEnum tipoDeJuros;
    private PeriodoDeJuros proximoPeriodo;
    private boolean fazendaPublica;
    private JurosEnum tabelaJuros;
    private BigDecimal meses;
    private BigDecimal taxa;

    public PeriodoDeJuros() {
    }

    public PeriodoDeJuros(Date dataInicial, Date dataFinal, BigDecimal aliquota, TipoDeQuantidadeDeJurosBaseEnum tipoDeQuantidade, TipoDeJurosEnum tipoDeJuros, boolean fazendaPublica, JurosEnum tabelaJuros) {
        this.dataInicial = dataInicial;
        this.dataFinal = dataFinal;
        this.aliquota = aliquota;
        this.tipoDeQuantidade = tipoDeQuantidade;
        this.tipoDeJuros = tipoDeJuros;
        this.fazendaPublica = fazendaPublica;
        this.tabelaJuros = tabelaJuros;
    }

    public PeriodoDeJuros clone() {
        return new PeriodoDeJuros(this.dataInicial, this.dataFinal, this.aliquota, this.tipoDeQuantidade, this.tipoDeJuros, this.fazendaPublica, this.tabelaJuros);
    }

    public Date getDataInicial() {
        return this.dataInicial;
    }

    public void setDataInicial(Date dataInicial) {
        this.dataInicial = dataInicial;
    }

    public Date getDataFinal() {
        return this.dataFinal;
    }

    public void setDataFinal(Date dataFinal) {
        this.dataFinal = dataFinal;
    }

    public BigDecimal getAliquota() {
        return this.aliquota;
    }

    public void setAliquota(BigDecimal aliquota) {
        this.aliquota = aliquota;
    }

    public TipoDeQuantidadeDeJurosBaseEnum getTipoDeQuantidade() {
        return this.tipoDeQuantidade;
    }

    public void setTipoDeQuantidade(TipoDeQuantidadeDeJurosBaseEnum tipoDeQuantidade) {
        this.tipoDeQuantidade = tipoDeQuantidade;
    }

    public TipoDeJurosEnum getTipoDeJuros() {
        return this.tipoDeJuros;
    }

    public void setTipoDeJuros(TipoDeJurosEnum tipoDeJuros) {
        this.tipoDeJuros = tipoDeJuros;
    }

    public PeriodoDeJuros getProximoPeriodo() {
        return this.proximoPeriodo;
    }

    public void setProximoPeriodo(PeriodoDeJuros proximoPeriodo) {
        this.proximoPeriodo = proximoPeriodo;
    }

    public long getDias() {
        return HelperDate.countDays(this.dataInicial, this.dataFinal) + 1L;
    }

    public BigDecimal getMeses() {
        if (this.meses == null) {
            HelperDate hdInicial = HelperDate.getInstance(this.dataInicial);
            HelperDate hdFinal = HelperDate.getInstance(this.dataFinal);
            this.meses = new BigDecimal(Long.valueOf(HelperDate.countMonths(this.dataInicial, this.dataFinal)).toString());
            long quantDias = HelperDate.countDays(this.dataInicial, hdInicial.lastDayOfTheMonth().getDate()) + 1L;
            if (TipoDeQuantidadeDeJurosBaseEnum.FRACAO.equals((Object)this.tipoDeQuantidade)) {
                BigDecimal fracao;
                int totalDiasMes = hdInicial.daysInMonth();
                if (quantDias < (long)totalDiasMes) {
                    fracao = new BigDecimal(quantDias).divide(new BigDecimal(totalDiasMes), Utils.CONTEXTO_MATEMATICO);
                    this.meses = this.meses.subtract(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO).add(fracao, Utils.CONTEXTO_MATEMATICO);
                }
                if ((quantDias = (long)hdFinal.getDay()) < (long)(totalDiasMes = hdFinal.daysInMonth())) {
                    fracao = new BigDecimal(quantDias).divide(new BigDecimal(totalDiasMes), Utils.CONTEXTO_MATEMATICO);
                    this.meses = this.meses.subtract(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO).add(fracao, Utils.CONTEXTO_MATEMATICO);
                }
            } else {
                if (quantDias < 15L) {
                    this.meses = this.meses.subtract(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO);
                }
                if (hdFinal.getDay() < 15) {
                    this.meses = this.meses.subtract(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO);
                }
            }
        }
        return this.meses;
    }

    public BigDecimal getTaxa() {
        if (this.taxa == null) {
            if (JurosEnum.JUROS_ZERO_TRINTA_TRES.equals((Object)this.getTabelaJuros())) {
                this.taxa = Utils.multiplicar(this.getAliquota(), new BigDecimal(this.getDias()));
            } else if (TipoDeJurosEnum.SIMPLES.equals((Object)this.tipoDeJuros)) {
                this.taxa = this.getAliquota().multiply(this.getMeses(), Utils.CONTEXTO_MATEMATICO);
            } else {
                this.taxa = BigDecimal.ONE.divide(new BigDecimal("100"), Utils.CONTEXTO_MATEMATICO).add(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO);
                this.taxa = new BigDecimal(Double.valueOf(Math.pow(this.taxa.doubleValue(), this.getMeses().doubleValue())).toString()).subtract(BigDecimal.ONE, Utils.CONTEXTO_MATEMATICO).multiply(new BigDecimal("100"), Utils.CONTEXTO_MATEMATICO);
            }
        }
        return this.taxa;
    }

    public boolean isFazendaPublica() {
        return this.fazendaPublica;
    }

    public void setFazendaPublica(boolean fazendaPublica) {
        this.fazendaPublica = fazendaPublica;
    }

    public JurosEnum getTabelaJuros() {
        return this.tabelaJuros;
    }

    public void setTabelaJuros(JurosEnum tabelaJuros) {
        this.tabelaJuros = tabelaJuros;
    }

    public String toString() {
        return String.format("[periodo: %s - %s, meses: %s, taxa: %s]", Utils.naoNulo(this.dataInicial) ? HelperDate.getInstance(this.dataInicial).format("dd/MM/yyyy") : "", Utils.naoNulo(this.dataFinal) ? HelperDate.getInstance(this.dataFinal).format("dd/MM/yyyy") : "", this.getMeses().toString(), this.getTaxa().toString());
    }
}

